### PR TITLE
[quant][pt2] Fix QAT conv-bn bias derived qspec

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112279
* __->__ #112159

Summary: Today, we have special handling for special qspecs like
`SharedQuantizationSpec` or `DerivedQuantizationSpec`, since these
qspecs refer to other nodes in the graph and these node references
need to be updated after replacement (since they referred to nodes
in the original graph that no longer exist in the new graph).

However, we only do the above for special nodes like conv, bn,
getitem, and relu. This doesn't cover the common use case of
having conv bias derive its qparams from those of conv input
activations and conv weight. This commit adds support for this
use case by also replacing the node references for these nodes.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_conv_bn_bias_derived_qspec

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50697078](https://our.internmc.facebook.com/intern/diff/D50697078)

### CODE REVIEW

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50697078).

File: torch/ao/quantization/pt2e/qat_utils.py [hunk 1]
```diff
@@ -407,8 +407,23 @@ def _get_nodes(nodes: List[Node]) -> Tuple[Node, Node, Node, Optional[Node]]:
...
+    assert isinstance(p_conv_weight, Node)
+    assert isinstance(r_conv_input, Node)
+    assert isinstance(r_conv_weight, Node)
+    o_conv_input = r.nodes_map[p_conv_input]
+    o_conv_weight = r.nodes_map[p_conv_weight]
+
     mapping = {
         "conv": (o_conv, r_conv),
+        "conv_input": (o_conv_input, r_conv_input),
+        "conv_weight": (o_conv_weight, r_conv_weight),
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/qat_utils.py:426]: add bias as well?

File: torch/ao/quantization/pt2e/qat_utils.py [hunk 2]
```diff
@@ -407,8 +407,23 @@ def _get_nodes(nodes: List[Node]) -> Tuple[Node, Node, Node, Optional[Node]]:
...
         "bn": (o_bn, r_bn),
         "getitem": (o_getitem, r_getitem),
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/qat_utils.py:428]: this is output?

Comment (andrewor14) [reply torch/ao/quantization/pt2e/qat_utils.py:428]: yeah, or the relu node below

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50697078).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50697078).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50697078).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 3]
```diff
@@ -608,6 +615,79 @@ def get_source_fn(node: torch.fx.Node):
...
+        example_inputs = (torch.randn(1, 3, 5, 5),)
+        m = capture_pre_autograd_graph(m, example_inputs)
+        quantizer = ConvBnDerivedBiasQuantizer()
+        m = prepare_qat_pt2e(m, quantizer)
+        m(*example_inputs)
+        m = convert_pt2e(m)
+        m(*example_inputs)
+
+
+class ConvBnDerivedBiasQuantizer(Quantizer):
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:629]: should we add a test for referring to bias node as well? e.g. qspec for weight being shared with bias

**Review by jerryzh168 — APPROVED** [review]
LG, maybe add a test for SharedQuantizationSpec of bias input edge as well

**Pull request closed without merging**